File: annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/policy/DefaultBehaviourProxyFactory.java
package com.github.dynamicextensionsalfresco.policy;

import java.util.ArrayList;

import com.github.dynamicextensionsalfresco.metrics.Timer;
import org.alfresco.repo.policy.Behaviour;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationListener;
import org.springframework.context.event.ContextClosedEvent;
/**
 * Created by jasper on 19/07/17.
 */
public class DefaultBehaviourProxyFactory  implements BehaviourProxyFactory, ApplicationListener<ContextClosedEvent> {
    private Logger logger = LoggerFactory.getLogger(DefaultBehaviourProxyFactory.class);

    @Autowired
    protected Timer timer;

    private ArrayList<BehaviourProxy> behaviourProxies = new ArrayList<BehaviourProxy>();

    @Override
    public BehaviourProxy createBehaviourProxy(Behaviour behaviour) {
        logger.debug("Creating BehaviourProxy for "+behaviour.getClass().getName()+"${} instance.");
        BehaviourProxy behaviourProxy = new BehaviourProxy(behaviour, timer);
        behaviourProxies.add(behaviourProxy);
        return behaviourProxy;
    }

    /* Callback operations */

    public void  onApplicationEvent( ContextClosedEvent event) {
        releaseBehaviourReferences();
    }

    /* Utility operations */

    protected void releaseBehaviourReferences() {
        for (BehaviourProxy behaviourProxy: behaviourProxies) {
            logger.debug("Releasing reference from BehaviourProxy to "+behaviourProxy.toString()+" instance." );
            behaviourProxy.release();
        }
        behaviourProxies.clear();
    }
}

File: annotations-runtime/src/main/kotlin/com/github/dynamicextensionsalfresco/policy/DefaultBehaviourProxyFactory.kt
//package com.github.dynamicextensionsalfresco.policy
//
//import com.github.dynamicextensionsalfresco.debug
//import java.util.ArrayList
//
//import com.github.dynamicextensionsalfresco.metrics.Timer
//import org.alfresco.repo.policy.Behaviour
//import org.slf4j.Logger
//import org.slf4j.LoggerFactory
//import org.springframework.beans.factory.annotation.Autowired
//import org.springframework.context.ApplicationListener
//import org.springframework.context.event.ContextClosedEvent
//
///**
// * Default [BehaviourProxyFactory] implementation intended for use in a Spring application context. This
// * implementation releases references to [Behaviour]s when the [ContextClosedEvent] occurs.
//
// * @author Laurens Fridael
// */
//public class DefaultBehaviourProxyFactoryOld @Autowired constructor(private val timer: Timer) : BehaviourProxyFactory, ApplicationListener<ContextClosedEvent> {
//
//    private val logger = LoggerFactory.getLogger(javaClass)
//
//    /* State */
//
//    private val behaviourProxies = ArrayList<BehaviourProxy>()
//
//    /* Main operations */
//
//    override fun createBehaviourProxy(behaviour: Behaviour): BehaviourProxy {
//        logger.debug {
//            "Creating BehaviourProxy for ${behaviour.javaClass.name} instance."
//        }
//        val behaviourProxy = BehaviourProxy(behaviour, timer)
//        behaviourProxies.add(behaviourProxy)
//        return behaviourProxy
//    }
//
//    /* Callback operations */
//
//    override fun onApplicationEvent(event: ContextClosedEvent) {
//        releaseBehaviourReferences()
//    }
//
//    /* Utility operations */
//
//    protected fun releaseBehaviourReferences() {
//        for (behaviourProxy in behaviourProxies) {
//            logger.debug { "Releasing reference from BehaviourProxy to $behaviourProxy instance." }
//            behaviourProxy.release()
//        }
//        behaviourProxies.clear()
//    }
//
//}
